README.md
# Bencoding
[![pub package](https://img.shields.io/pub/v/bencoding.svg)](https://pub.dev/packages/bencoding)

A dart library for encoding and decoding bencoded data,
according to the [BitTorrent specification](https://wiki.theory.org/index.php/BitTorrentSpecification#Bencoding).

## Index

- [About BEncoding](#about-bencoding)
- [Usage](#usage)
- [API](#api)

## About BEncoding

from [Wikipedia](https://en.wikipedia.org/wiki/Bencode):

Bencode (pronounced like B encode) is the encoding used by the peer-to-peer
file sharing system BitTorrent for storing and transmitting loosely structured data.

It supports four different types of values:
- byte strings
- integers
- lists
- dictionaries

Bencoding is most commonly used in torrent files.
These metadata files are simply bencoded dictionaries.

## Usage

### Import
```dart
import 'package:bencoding/bencode.dart' as bencode;
```

### Encoding

```dart
final data = {
  'string': 'Hello World',
  'integer': 12345,
  'dict': {
    'key': 'This is a string within a dictionary',
  },
  'list': [ 1, 2, 3, 4, 'string', 5, {} ],
};

final buffer = bencode.encode(data); // Return a Uint8List
final string = bencode.encodeToString(data); // Return a String

print('Buffer: $buffer');
print('String: $string');
```

#### Output

```
Buffer: [100, 54, 58, 115, 116, 114, 105, 110, 103, 49, 49, ...]
String: d6:string11:Hello World7:integeri12345e4:dictd3:key36:This is ...
```

### Decoding

```dart
final bencodedData = 'd6:string11:Hello World7:integeri12345e4:dictd3:key36:This is a string within a dictionarye4:listli1ei2ei3ei4e6:stringi5edeee';

print(bencode.decodeString(bencodedData));
// or you can use bencode.deocode(data) if your input data is a buffer (Uint8List)
```

#### Output

```dart
{
  'string': 'Hello World',
  'integer': 12345,
  'dict': {
    'key': 'This is a string within a dictionary',
  },
  'list': [ 1, 2, 3, 4, 'string', 5, {}],
}
```

## API

### encode( `dynamic` *data* )
> `String` | `int` | `List` | `Map` *data*

Returns `Uint8List`

### encodeToString ( `dynamic` *data* )
> `String` | `int` | `List` | `Map` *data*

Returns `String`

### decode( `Uint8List` *data* )
Returns `dynamic` (`String` | `int` | `List` | `Map`)

### decodeString ( `String` *data* )
Returns `dynamic` (`String` | `int` | `List` | `Map`)


CHANGELOG.md
## 1.0.0

- Now the decoder check if strings are utf8 chars only, if is not the case the result will be a Uint8List

## 0.1.2

- Updating README.md

## 0.1.1

- Cleaning & formating

## 0.1.0

- First version with every based features


lib/src/encoder.dart
import 'dart:convert';
import 'dart:typed_data';

/// Encodes data in bencode.
Uint8List encode(dynamic data) {
  List<Uint8List> buffers = [];

  _addToBuffers(buffers, data);

  return Uint8List.fromList(buffers.expand((x) => x).toList());
}

/// Encodes data in bencode String.
String encodeToString(dynamic data) {
  return utf8.decode(encode(data));
}

Uint8List _string(String data) {
  return utf8.encode('${data.length}:$data');
}

Uint8List _integer(int data) {
  return utf8.encode('i${data}e');
}

List<Uint8List> _list(List data) {
  List<Uint8List> buffers = [
    Uint8List.fromList([108])
  ];

  for (var item in data) {
    if (item == null) continue;

    _addToBuffers(buffers, item);
  }

  buffers.add(Uint8List.fromList([101]));

  return buffers;
}

List<Uint8List> _dictionary(Map data) {
  List<Uint8List> buffers = [
    Uint8List.fromList([100])
  ];

  data.forEach((key, value) {
    if (value == null) return;

    buffers.add(_string(key));

    _addToBuffers(buffers, value);
  });

  buffers.add(Uint8List.fromList([101]));

  return buffers;
}

_addToBuffers(List<Uint8List> buffers, dynamic data) {
  if (data is String) {
    buffers.add(_string(data));
  } else if (data is int) {
    buffers.add(_integer(data));
  } else if (data is bool) {
    buffers.add(_integer(data ? 1 : 0));
  } else if (data is List) {
    buffers.addAll(_list(data));
  } else if (data is Map) {
    buffers.addAll(_dictionary(data));
  } else {
    throw TypeError();
  }
}


lib/src/decoder.dart
import 'dart:convert';
import 'dart:typed_data';

/// Decode bencoded data.
dynamic decode(Uint8List data) {
  var index = _Index(0);

  while (data[index.value] == 32 || data[index.value] == 11 || data[index.value] == 9) {
    index.value++;
  }

  return _getNextData(data, index);
}

/// Decode bencoded data from a String.
dynamic decodeString(String data) {
  return decode(utf8.encode(data));
}

Map<String, dynamic> _dictionary(Uint8List data, _Index index) {
  Map<String, dynamic> result = Map();

  index.value++;

  while (data[index.value] != 101) {
    result[utf8.decode(_bytes(data, index))] = _getNextData(data, index);
  }

  index.value++;

  return result;
}

List<dynamic> _list(Uint8List data, _Index index) {
  List<dynamic> result = [];

  index.value++;

  while (data[index.value] != 101) {
    result.add(_getNextData(data, index));
  }

  index.value++;

  return result;
}

int _integer(Uint8List data, _Index index) {
  final end = data.indexOf(101, index.value);
  final result = _parseInt(data, index.value + 1, end);

  index.value = end + 1;

  return result;
}

Uint8List _bytes(Uint8List data, _Index index) {
  final sep = data.indexOf(58, index.value);
  final length = _parseInt(data, index.value, sep);
  final bytes = data.sublist(sep + 1, sep + 1 + length);

  index.value = sep + 1 + length;

  return bytes;
}

int _parseInt(Uint8List data, int start, int end) {
  var sum = 0;
  var sign = 1;

  for (var i = start; i < end; i++) {
    var number = data[i];

    if (number < 58 && number >= 48) {
      sum = sum * 10 + (number - 48);
    } else if (i == start && number == 43) {
      continue;
    } else if (i == start && number == 45) {
      sign = -1;
    } else if (number == 46) {
      break;
    } else {
      throw TypeError();
    }
  }

  return sum * sign;
}

dynamic _getNextData(Uint8List data, _Index index) {
  if (data[index.value] == 100) {
    return _dictionary(data, index);
  } else if (data[index.value] == 108) {
    return _list(data, index);
  } else if (data[index.value] == 105) {
    return _integer(data, index);
  } else {
    final bytes = _bytes(data, index);

    try {
      return utf8.decode(bytes);
    } on FormatException {
      return bytes;
    }
  }
}

class _Index {
  int value;
  _Index(this.value);
}
